### PR TITLE
Allow appending and prepending copy-from descriptions

### PR DESCRIPTION
**Pull request by Procyonae**
#### Summary
None

#### Purpose of change

Aid for #71112 and #70352
Less repetition -> less translation needed + less typo opportunities + less bloat -> yay

#### Describe the solution

Adds optional "description_prepend" and "description_append" to items and item variants that add text to the start or end of the description respectively. Combined with copy-from this allows adding to descriptions without repetition.
Adds a disgustingly named (please suggest alternatives) "inherit_extended_description" bool that defaults to true but can be specified to false in which case items that inherit from that item will only use the "description" not the extended description after prepending/appending/both, useful for stuff like steel grades where the base item can append its grade without it being inherited to save adding abstract entries for all steel graded stuff.
The old "append" behaviour for item variants is still supported to avoid making this PR harder to review, will remove in a follow-up PR.

#### Describe alternatives you've considered

Adding better context to translator comments but that would require searching through the json to find information about copy-from members which is probably expensive even without doing it recursively and only checking the same file
I've no idea if extended_description() is defined in a sensible place nor whether everything should be public
Idk if there's much performance benefit if any to having the extend_description bool

#### Testing
Tested with custom test stuff like below and the changes in the PR
```
  {
    "id": "magic_1_ball",
    "type": "GENERIC",
    "category": "other",
    "name": { "str": "Magic 1-Ball" },
    "description": "Once I were a 1-Ball.",
    ...
  },
  {
    "id": "magic_2_ball",
    "type": "GENERIC",
    "copy-from": "magic_1_ball",
    "name": { "str": "Magic 2-Ball" },
    "description_prepend": "Twice a 2-Ball.",
    "description_append": "Twice a 2-Ball."
  },
  {
    "id": "magic_3_ball",
    "type": "GENERIC",
    "copy-from": "magic_2_ball",
    "name": { "str": "Magic 3-Ball" },
    "description_append": "I even became a 3-Ball."
  },
  {
    "id": "magic_4_ball",
    "type": "GENERIC",
    "copy-from": "magic_3_ball",
    "name": { "str": "Magic 4-Ball" },
    "description_prepend": "I came to be a 4-Ball.",
    "description_append": "But decided a 4-Ball was my true form."
  },
```

image

### CODE REVIEW

**Comment by GuardianDll** [comment]
btw, `name` and `description` is still mandatory for items, and not carried by inheritance

**Comment by Procyonae** [comment]
> btw, `name` and `description` is still mandatory for items, and not carried by inheritance

Description isn't the test example above literally shows it isn't?

**Comment by GuardianDll** [comment]
Huh, i was sure description is also mandatory

**Comment by Procyonae** [comment]
@BrettDong I could use a review from someone who understands how the translation stuff works as to whether I'm creating unnecessary translation strings or missing necessary ones anywhere.

File: lang/string_extractor/parsers/generic.py [hunk 1]
```diff
@@ -35,11 +56,20 @@ def parse_generic(json, origin):
...
             variant_name = get_singular_name(variant["name"])
             write_text(variant["name"], origin,
-                       comment="Variant name of item \"{}\"".format(name),
-                       plural=True)
-            write_text(variant["description"], origin,
-                       comment="Description of variant \"{1}\" of item \"{0}\""
-                       .format(name, variant_name))
+                    comment="Variant name of item \"{}\"".format(name),
+                    plural=True)
+            if "description" in json["variants":
```
Comment (Procyonae) [inline lang/string_extractor/parsers/generic.py:61]:
```suggestion
            if "description" in json["variants"]:
```

File: lang/string_extractor/parsers/generic.py [hunk 2]
```diff
@@ -35,11 +56,20 @@ def parse_generic(json, origin):
...
+                write_text(variant["description"], origin,
+                        comment="Description of variant \"{}\" of item \"{}\""
+                        .format(variant_name, name))
+            if "description_prepend" in json["variants":
```
Comment (Procyonae) [inline lang/string_extractor/parsers/generic.py:65]:
```suggestion
            if "description_prepend" in json["variants"]:
```

File: lang/string_extractor/parsers/generic.py [hunk 3]
```diff
@@ -35,11 +56,20 @@ def parse_generic(json, origin):
...
+                write_text(variant["description_prepend"], origin,
+                        comment="Partial description of variant \"{}\" of item \"{}\" to add to the start of \"{}\""
+                        .format(variant_name, name, description))
+            if "description_append" in json["variants":
```
Comment (Procyonae) [inline lang/string_extractor/parsers/generic.py:69]:
```suggestion
            if "description_append" in json["variants"]:
```

File: lang/string_extractor/parsers/generic.py [hunk 4]
```diff
@@ -21,6 +26,22 @@ def parse_generic(json, origin):
     if "description" in json:
         write_text(json["description"], origin, c_format=False,
                    comment=comment + ["Description of \"{}\"".format(name)])
+        description = json["description"]
+        inherits_description = false
+    if "description_prepend" in json:
+        if inherits_description:
+            write_text(json["description_prepend"], origin, c_format=False,
+                    comment=comment + ["Partial description of \"{}\" to add to the start of its inherited description from \"{}\"".format(name, inherited_description_id)])
```
Comment (Procyonae) [inline lang/string_extractor/parsers/generic.py:34]:
```suggestion
            write_text(json["description_prepend"], origin, c_format=False,
                       comment=comment + ["Partial description of \"{}\" to 
                       add to the start of its inherited description from 
                       \"{}\"".format(name, inherited_description_id)])
```

File: lang/string_extractor/parsers/generic.py [hunk 5]
```diff
@@ -21,6 +26,22 @@ def parse_generic(json, origin):
...
+        else:
+            write_text(json["description_prepend"], origin, c_format=False,
+                    comment=comment + ["Partial description of \"{}\" to add to the start of \"{}\"".format(name, description)])
```
Comment (Procyonae) [inline lang/string_extractor/parsers/generic.py:37]:
```suggestion
            write_text(json["description_prepend"], origin, c_format=False,
                       comment=comment + ["Partial description of \"{}\" to 
                       add to the start of \"{}\"".format(name, description)])
```

File: lang/string_extractor/parsers/generic.py [hunk 6]
```diff
@@ -21,6 +26,22 @@ def parse_generic(json, origin):
...
+    if "description_append" in json:
+        if inherits_description:
+            write_text(json["description_append"], origin, c_format=False,
+                    comment=comment + ["Partial description of \"{}\" to add to the end of its inherited description from \"{}\"".format(name, inherited_description_id)])
```
Comment (Procyonae) [inline lang/string_extractor/parsers/generic.py:41]:
```suggestion
            write_text(json["description_append"], origin, c_format=False,
                       comment=comment + ["Partial description of \"{}\" to 
                       add to the end of its inherited description from 
                       \"{}\"".format(name, inherited_description_id)])
```

File: lang/string_extractor/parsers/generic.py [hunk 7]
```diff
@@ -21,6 +26,22 @@ def parse_generic(json, origin):
...
+        else:
+            write_text(json["description_append"], origin, c_format=False,
+                    comment=comment + ["Partial description of \"{}\" to add to the end of \"{}\"".format(name, description)])
```
Comment (Procyonae) [inline lang/string_extractor/parsers/generic.py:44]:
```suggestion
            write_text(json["description_append"], origin, c_format=False,
                       comment=comment + ["Partial description of \"{}\" to 
                       add to the end of \"{}\"".format(name, description)])
```

File: lang/string_extractor/parsers/generic.py [hunk 8]
```diff
@@ -21,6 +26,28 @@ def parse_generic(json, origin):
...
+    if "description_append" in json:
+        if inherits_description:
+            write_text(json["description_append"], origin, c_format=False,
+                       comment=comment + ["Partial description of \"{}\" to 
+                       add to the end of its inherited description from 
+                       \"{}\"".format(name, inherited_description_id)])
+        else:
+            write_text(json["description_append"], origin, c_format=False,
+                       comment=comment + ["Partial description of \"{}\" to 
+                       add to the end of \"{}\"".format(name, description)])
```
Comment (Procyonae) [inline lang/string_extractor/parsers/generic.py:50]:
```suggestion
            write_text(json["description_prepend"], origin, c_format=False,
                       comment=comment + ["Partial description of \"{}\" to add to the start of its inherited description from \"{}\""
                       .format(name, inherited_description_id)])
        else:
            write_text(json["description_prepend"], origin, c_format=False,
                       comment=comment + ["Partial description of \"{}\" to add to the start of \"{}\""
                       .format(name, description)])
    if "description_append" in json:
        if inherits_description:
            write_text(json["description_append"], origin, c_format=False,
                       comment=comment + ["Partial description of \"{}\" to add to the end of its inherited description from \"{}\""
                       .format(name, inherited_description_id)])
        else:
            write_text(json["description_append"], origin, c_format=False,
                       comment=comment + ["Partial description of \"{}\" to add to the end of \"{}\""
                       .format(name, description)])
```

**Comment by Procyonae** [comment]
Kevin is debating whether this should go through or whether the existing append feature from #56302 should be scrapped  so this'll be drafted until they make a decision, see [this message chain on the devcord](https://discord.com/channels/598523535169945603/598529174302490644/1200587905576468621)

File: lang/string_extractor/parsers/generic.py [hunk 9]
```diff
@@ -21,6 +26,26 @@ def parse_generic(json, origin):
...
+                       comment=comment + ["Partial description of \"{}\" to add to the start of \"{}\""
+                       .format(name, description)])
+    if "description_append" in json:
+        if inherits_description:
+            write_text(json["description_append"], origin, c_format=False,
+                       comment=comment + ["Partial description of \"{}\" to add to the end of its inherited description from \"{}\""
+                       .format(name, inherited_description_id)])
+        else:
+            write_text(json["description_append"], origin, c_format=False,
+                       comment=comment + ["Partial description of \"{}\" to add to the end of \"{}\""
```
Comment (Procyonae) [inline lang/string_extractor/parsers/generic.py:47]:
```suggestion
        description = json["description"]
        inherits_description = False
    if "description_prepend" in json:
        if inherits_description:
            write_text(json["description_prepend"], origin, c_format=False,
                       comment=comment +
                       ["Partial description of \"{}\" to add to the start of its inherited description from \"{}\""
                       .format(name, inherited_description_id)])
        else:
            write_text(json["description_prepend"], origin, c_format=False,
                       comment=comment +
                       ["Partial description of \"{}\" to add to the start of \"{}\""
                       .format(name, description)])
    if "description_append" in json:
        if inherits_description:
            write_text(json["description_append"], origin, c_format=False,
                       comment=comment +
                       ["Partial description of \"{}\" to add to the end of its inherited description from \"{}\""
                       .format(name, inherited_description_id)])
        else:
            write_text(json["description_append"], origin, c_format=False,
                       comment=comment +
                       ["Partial description of \"{}\" to add to the end of \"{}\""
```

File: lang/string_extractor/parsers/generic.py [hunk 10]
```diff
@@ -35,11 +60,20 @@ def parse_generic(json, origin):
...
-                       .format(name, variant_name))
+                    comment="Variant name of item \"{}\"".format(name),
+                    plural=True)
+            if "description" in json["variants"]:
+                write_text(variant["description"], origin,
+                        comment="Description of variant \"{}\" of item \"{}\""
+                        .format(variant_name, name))
+            if "description_prepend" in json["variants"]:
+                write_text(variant["description_prepend"], origin,
+                        comment="Partial description of variant \"{}\" of item \"{}\" to add to the start of \"{}\""
```
Comment (Procyonae) [inline lang/string_extractor/parsers/generic.py:71]:
```suggestion
                       comment="Variant name of item \"{}\"".format(name),
                       plural=True)
            if "description" in json["variants"]:
                write_text(variant["description"], origin, comment=
                           "Description of variant \"{}\" of item \"{}\""
                           .format(variant_name, name))
            if "description_prepend" in json["variants"]:
                write_text(variant["description_prepend"], origin, comment=
                           "Partial description of variant \"{}\" of item \"{}\" to add to the start of \"{}\""
```

**Review by Procyonae — COMMENTED** [review]
Srsly who thought 79 characters a line was enough

File: lang/string_extractor/parsers/generic.py [hunk 11]
```diff
@@ -21,6 +26,30 @@ def parse_generic(json, origin):
...
+    if "description_append" in json:
+        if inherits_description:
+            write_text(json["description_append"], origin, c_format=False,
+                       comment=comment +
+                       ["Partial description of \"{}\" to add to the end of its inherited description from \"{}\""
+                       .format(name, inherited_description_id)])
+        else:
+            write_text(json["description_append"], origin, c_format=False,
+                       comment=comment +
+                       ["Partial description of \"{}\" to add to the end of \"{}\""
```
Comment (Procyonae) [inline lang/string_extractor/parsers/generic.py:51]:
```suggestion
                       comment=comment + ["Partial description of \"{}\" to "
                       "add to the start of its inherited description from \"{}\""
                       .format(name, inherited_description_id)])
        else:
            write_text(json["description_prepend"], origin, c_format=False,
                       comment=comment + ["Partial description of \"{}\" to "
                       "add to the start of \"{}\"".format(name, description)])
    if "description_append" in json:
        if inherits_description:
            write_text(json["description_append"], origin, c_format=False,
                       comment=comment + ["Partial description of \"{}\" to "
                       "add to the end of its inherited description from \"{}\""
                       .format(name, inherited_description_id)])
        else:
            write_text(json["description_append"], origin, c_format=False,
                       comment=comment + ["Partial description of \"{}\" to "
                       "add to the end of \"{}\""
```

File: lang/string_extractor/parsers/generic.py [hunk 12]
```diff
@@ -37,9 +66,18 @@ def parse_generic(json, origin):
...
+                           "Description of variant \"{}\" of item \"{}\""
+                           .format(variant_name, name))
+            if "description_prepend" in json["variants"]:
+                write_text(variant["description_prepend"], origin, comment=
+                           "Partial description of variant \"{}\" of item \"{}\" to add to the start of \"{}\""
+                        .format(variant_name, name, description))
+            if "description_append" in json["variants"]:
+                write_text(variant["description_append"], origin,
+                        comment="Partial description of variant \"{}\" of item \"{}\" to add to the end of \"{}\""
+                        .format(variant_name, name, description))
```
Comment (Procyonae) [inline lang/string_extractor/parsers/generic.py:80]:
```suggestion
                write_text(variant["description"], origin,
                           comment="Description of variant \"{}\" of item "
                           "\"{}\"".format(variant_name, name))
            if "description_prepend" in json["variants"]:
                write_text(variant["description_prepend"], origin,
                           comment="Partial description of variant \"{}\" of "
                           "item \"{}\" to add to the start of \"{}\""
                           .format(variant_name, name, description))
            if "description_append" in json["variants"]:
                write_text(variant["description_append"], origin,
                           comment="Partial description of variant \"{}\" of "
                           "item \"{}\" to add to the end of \"{}\""
                           .format(variant_name, name, description))
```

File: lang/string_extractor/parsers/generic.py [hunk 13]
```diff
@@ -21,6 +26,29 @@ def parse_generic(json, origin):
...
+        if inherits_description:
+            write_text(json["description_append"], origin, c_format=False,
+                       comment=comment + ["Partial description of \"{}\" to "
+                       "add to the end of its inherited description from \"{}\""
+                       .format(name, inherited_description_id)])
+        else:
+            write_text(json["description_append"], origin, c_format=False,
+                       comment=comment + ["Partial description of \"{}\" to "
+                       "add to the end of \"{}\""
+                       .format(name, description)])
```
Comment (Procyonae) [inline lang/string_extractor/parsers/generic.py:51]:
```suggestion
            write_text(json["description_prepend"], origin, c_format=False,
                       comment=comment + ["Partial description of \"{}\" to "
                       "add to the start of its inherited description from "
                       "\"{}\"".format(name, inherited_description_id)])
        else:
            write_text(json["description_prepend"], origin, c_format=False,
                       comment=comment + ["Partial description of \"{}\" to "
                       "add to the start of \"{}\"".format(name, description)])
    if "description_append" in json:
        if inherits_description:
            write_text(json["description_append"], origin, c_format=False,
                       comment=comment + ["Partial description of \"{}\" to "
                       "add to the end of its inherited description from "
                       "\"{}\"".format(name, inherited_description_id)])
        else:
            write_text(json["description_append"], origin, c_format=False,
                       comment=comment + ["Partial description of \"{}\" to "
                       "add to the end of \"{}\"".format(name, description)])
```

**Comment by BrettDong** [comment]
> whether I'm creating unnecessary translation strings or missing necessary ones anywhere.

You can check out the output in "List text changes" step in "Text Changes Analyser" CI job to see the impact on translatable strings from the pull request:

https://github.com/CleverRaven/Cataclysm-DDA/actions/runs/7678085991/job/20927556843?pr=71121#step:9:13

It lists 1556 strings deleted and only 7 added, so looks like a lot of stuff is missing.

**Comment by BrettDong** [comment]
If you need help on Python code formatting: I use `autopep8` to format Python code:

```
cd ./lang/string_extractor
autopep8
```

File: lang/string_extractor/parsers/generic.py [hunk 14]
```diff
@@ -21,6 +26,28 @@ def parse_generic(json, origin):
...
         write_text(json["description"], origin, c_format=False,
                    comment=comment + ["Description of \"{}\"".format(name)])
+        description = json["description"]
+        inherits_description = False
+    if "description_prepend" in json:
+        if inherits_description:
+            write_text(json["description_prepend"], origin, c_format=False,
+                       comment=comment + ["Partial description of \"{}\" to "
+                       "add to the start of its inherited description from "
+                       "\"{}\"".format(name, inherited_description_id)])
```
Comment (Procyonae) [inline lang/string_extractor/parsers/generic.py:36]:
```suggestion
                                          "\"{}\"".format(name, inherited_description_id)])
```

File: lang/string_extractor/parsers/generic.py [hunk 15]
```diff
@@ -21,6 +26,28 @@ def parse_generic(json, origin):
...
+        else:
+            write_text(json["description_prepend"], origin, c_format=False,
+                       comment=comment + ["Partial description of \"{}\" to "
+                       "add to the start of \"{}\"".format(name, description)])
+    if "description_append" in json:
+        if inherits_description:
+            write_text(json["description_append"], origin, c_format=False,
+                       comment=comment + ["Partial description of \"{}\" to "
+                       "add to the end of its inherited description from "
+                       "\"{}\"".format(name, inherited_description_id)])
```
Comment (Procyonae) [inline lang/string_extractor/parsers/generic.py:46]:
```suggestion
                                          "\"{}\"".format(name, inherited_description_id)])
```

File: lang/string_extractor/parsers/generic.py [hunk 16]
```diff
@@ -37,9 +64,20 @@ def parse_generic(json, origin):
...
+            if "description_prepend" in json["variants"]:
+                write_text(variant["description_prepend"], origin,
+                           comment="Partial description of variant \"{}\" of "
+                           "item \"{}\" to add to the start of \"{}\""
+                           .format(variant_name, name, description))
+            if "description_append" in json["variants"]:
+                write_text(variant["description_append"], origin,
+                           comment="Partial description of variant \"{}\" of "
+                           "item \"{}\" to add to the end of \"{}\""
+                           .format(variant_name, name, description))
```
Comment (Procyonae) [inline lang/string_extractor/parsers/generic.py:80]:
```suggestion
            if "description" in variant:
                write_text(variant["description"], origin,
                           comment="Description of variant \"{}\" of item "
                           "\"{}\"".format(variant_name, name))
            if "description_prepend" in variant:
                write_text(variant["description_prepend"], origin,
                           comment="Partial description of variant \"{}\" of "
                           "item \"{}\" to add to the start of \"{}\""
                           .format(variant_name, name, description))
            if "description_append" in variant:
                write_text(variant["description_append"], origin,
                           comment="Partial description of variant \"{}\" of "
                           "item \"{}\" to add to the end of \"{}\""
                           .format(variant_name, name, description))
```

File: src/item.cpp [hunk 17]
```diff
@@ -9278,16 +9278,24 @@ void item::set_itype_variant( const std::string &variant )
...
 {
     if( !has_itype_variant() ) {
-        return "";
+        return ""; // Shouldn't this throw an error?
     }
-
-    // append the description instead of fully overwriting it
-    if( itype_variant().append ) {
-        return _( string_format( "%s  %s", type->description.translated(),
-                                 itype_variant().alt_description.translated() ) );
```
Comment (Procyonae) [inline src/item.cpp:9287]: @BrettDong Was there any point to this return value being exposed to translation? The text changes thing you pointed out doesn't say anything unexpected is removed now so I'm guessing they weren't getting exposed anyway? Or do the C++ translation changes show up separately?

Comment (BrettDong) [reply src/item.cpp:9287]: I think this is a code mistake. No point in calling _() on the return value from string format.

**Comment by RenechCDDA** [comment]
Reopened on request of the author.

**Comment by Procyonae** [comment]
Having a bit of a mare with this one. After further discussion with kevin this is fine to add as long as the full extended string is exposed to translation as opposed to the individual sentences which felt like it should be easy enough but bc of the base description potentially being copy-frommed multiple times the python code we use to add comments to it currently has no simple way to piece it together and I have no idea how to do dynamic C++ translation comments if we even can.

**Comment by Uwuewsky** [comment]
From my experience, I think it would be enough to write in the comment that this line is added to the description of the specified thing. If anyone is unsure of the context, they can look it up.

Yes, it will not be easy to write down in the comments the entire description of what it will be like ingame. An item can be copied from an item in another file.

We can try, as far as I've seen, it's rare that a definition is copied more than one level of inheritance. Perhaps it is possible in most cases to write down a complete description, otherwise fallback to specifying the copied entity.

**Comment by GuardianDll** [comment]
should this be reopen, or put into "abandoned but desired PRs" list, or else?

**Comment by Procyonae** [comment]
I'm not working on it rn (and I can reopen stuff myself now) but it's a bit of a mess so I wouldn't bother adding it to "abandoned but desired PRs", idm if it is tho

**Pull request closed without merging**